xapian-applications/omega/query.cc

```diff
@@ -752,11 +756,13 @@ html_highlight(const string &s, const string &list,
 	}
 	j = term_end;
 	term = Xapian::Unicode::tolower(term);
-	int match = word_in_list(term, list);
+	build_word_map(list);
```

Comment:
This is the wrong place to build the mapping - you're doing it for every word in the text being highlighted.

You want to do it at most once for every call to `$highlight` (and it could be reused if `$highlight` is called again with the same list of terms to highlight, which is quite a likely case - e.g. highlighting matches in the document title and a sample of text from the document).